runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/ClockSkewInterceptor.kt

```diff
@@ -0,0 +1,96 @@
+import aws.smithy.kotlin.runtime.telemetry.logging.logger
+import aws.smithy.kotlin.runtime.time.Instant
+import aws.smithy.kotlin.runtime.time.until
+import kotlinx.atomicfu.*
+import kotlin.coroutines.coroutineContext
+import kotlin.time.Duration
+import kotlin.time.Duration.Companion.minutes
+
+/**
+ * An interceptor used to detect clock skew (difference between client and server clocks) and apply a correction.
+ */
+public class ClockSkewInterceptor : HttpInterceptor {
+    public companion object {
+        /**
+         * How much must the clock be skewed before attempting correction
+         */
+        public val CLOCK_SKEW_THRESHOLD: Duration = 4.minutes
+
+        /**
+         * Get the current skew between the client time and [serverTime] as a [Duration].
+         * It may be negative if the serverTime is in the past.
+         * @param serverTime the server's time
+         */
+        internal fun Instant.getSkew(serverTime: Instant): Duration = this.until(serverTime)
+
+        /**
+         * Determine whether the client's clock is skewed relative to the server.
+         * @param serverTime the server's time
+         */
+        internal fun Instant.isSkewed(serverTime: Instant): Boolean = getSkew(serverTime).absoluteValue >= CLOCK_SKEW_THRESHOLD
+    }
+
+    // Clock skew to be applied to all requests
+    private val _currentSkew: AtomicRef<Duration> = atomic(Duration.ZERO)
+
+    private val currentSkew: Duration
+        get() = _currentSkew.value
+
+    /**
+     * Apply the previously-computed skew, if it's set, to the execution context before signing
+     */
+    public override suspend fun modifyBeforeSigning(context: ProtocolRequestInterceptorContext<Any, HttpRequest>): HttpRequest {
+        val logger = coroutineContext.logger<ClockSkewInterceptor>()
+
+        val skew = currentSkew
+        if (skew != Duration.ZERO) {
+            logger.debug { "applying clock skew $skew to client" }
+            context.executionContext[HttpOperationContext.ClockSkew] = skew
+        }
+
+        return context.protocolRequest
+    }
+
+    /**
+     * After receiving a response, check if the client clock is skewed and apply a correction if necessary.
+     */
+    public override suspend fun modifyBeforeDeserialization(context: ProtocolResponseInterceptorContext<Any, HttpRequest, HttpResponse>): HttpResponse {
+        val logger = coroutineContext.logger<ClockSkewInterceptor>()
+
+        val serverTime = context.protocolResponse.header("Date")?.let {
+            Instant.fromRfc5322(it)
+        } ?: run {
+            logger.info { "service did not return \"Date\" header, skipping skew calculation" }
+            return context.protocolResponse
+        }
+
+        val clientTime = context.protocolRequest.headers["Date"]?.let {
+            Instant.fromRfc5322(it)
+        } ?: context.protocolRequest.headers["x-amz-date"]?.let {
+            Instant.fromIso8601(it)
+        } ?: Instant.now()
```

Comment:
**Correctness**: Defaulting to `Instant.now()` seems wrong. Since there are a bunch of suspension points around receiving, it could be many minutes later when we _process_ a response vs when the response _actually arrived_ at the client. This might calculate a client-local time that's vastly different than when the request was first serialized.

If we truly do want a fallback in case of no outgoing `Date`/`x-amz-date` header, I suggest we put it in the execution context during `modifyBeforeSigning`.